server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -242,59 +262,93 @@ public Object persist(final Committer committer) throws InterruptedException
   }
 
   /**
-   * Publish all data indexed through this driver so far, and waits for it to be handed off. Blocks until complete.
-   * Retries forever on transient failures, but may exit early on permanent failures.
-   *
-   * Should be called after all data has been added and persisted through {@link #add(InputRow, String, Supplier)} and
-   * {@link #persist(Committer)}.
+   * Register the segments in the given {@link SegmentsAndMetadata} to be handed off and execute a background task which
+   * waits until the hand off completes.
    *
-   * @param publisher publisher to use for this set of segments
-   * @param committer committer representing all data that has been added so far
-   *
-   * @return segments and metadata published if successful, or null if segments could not be handed off due to
-   * transaction failure with commit metadata.
+   * @param segmentsAndMetadata
+   * @return
+   * @throws InterruptedException
    */
-  public SegmentsAndMetadata finish(
-      final TransactionalSegmentPublisher publisher,
-      final Committer committer
-  ) throws InterruptedException
+  public ListenableFuture<SegmentsAndMetadata> registerHandoff(SegmentsAndMetadata segmentsAndMetadata)
```

Comment:
> I see a potential problem here. If handoffConditionTimeout is set and publishes are happening in another executor in a way that on publish failure it is being retired infinitely then even though no publish succeeded because of the handoffConditionTimeout a task may finish successfully.

@pjain1 I don't understand your comment. Would you elaborate more? Also, I deprecated `handoffConditionTimeout`. Please consider it too.